### PR TITLE
BETA version 2 of the Storybook Multi Input (Editing and Accessibility)

### PR DESCRIPTION
**Pull request by Ridhwana**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
In the [previous PR ](https://github.com/forem/forem/pull/17828), I worked on creating the ability to add items to a field. 

In this PR, I tackle:
1.  Editing the selected items
2. Resizing the input when editing a selected item 
3. Allowing the use of a shared Default Selection component between the multi autocomplete component and the multi input component with additional customizable props.
4. Passing through the input regex as a prop
5. Deleting and editing using keyboard navigation (accessibility) 

I'm working on:
- validation
- tests
- documentation
in separate PRs.

## Related Tickets & Documents

- Related Issue #https://github.com/forem/forem/issues/17577

## QA Instructions, Screenshots, Recordings

- Locally run yarn install && yarn storybook
- Navigate to the BETA > MultiInput story
- Play around with editing items (email addresses).

In this demo, I show the editing of a component and an accessibility demo

https://www.loom.com/share/14e02343b49647f6a6c2f3c4f38eeebe

### UI accessibility concerns?

I tested keyboard navigation and voice over, see loom above. 

## Added/updated tests?

- [ ] Yes
- [x] No, and this is why: This will come next. I think it's okay for now, since it's in BETA.
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

Some notes from [previous PR](https://github.com/forem/forem/pull/17828): 

> In terms of a11y testing, I think the main thing missing is a confirmation announcement when an item is added or removed from the list. In the MultiSelectAutocomplete we achieve that with a hidden aria-live region, but depending whether or not we are adding the "Edit in place" functionality to this feature, there could potentially be a simpler path forward.
> 
> There's also an issue at the moment with keyboard access to the pill items, as I can tab to them but not press Enter for the delete action (as the click handler is on the SVG rather than the button). Again, depending on whether we add the "edit" functionality, we'll need to make some changes to the selected items markup.
> 
> These things can definitely be tackled in subsequent PRs - just making sure we capture them as tasks!

_____

### CODE REVIEW

File: app/javascript/crayons/MultiInput/__stories__/MultiInput.stories.jsx [hunk 1]
```diff
@@ -14,6 +14,9 @@ export default {
         description:
           'Placeholder text, shown when no selections have been made yet',
       },
+      regex: {
+        description: 'A regular expression to used to validate the input',
```
[the remark this answers is not in the document]
Comment (Ridhwana) [reply app/javascript/crayons/MultiInput/__stories__/MultiInput.stories.jsx:18]: Oops, thats one to many "to's" - thanks for catching this.

File: app/javascript/crayons/MultiInput/MultiInput.jsx [hunk 2]
```diff
@@ -17,90 +17,219 @@ const ALLOWED_CHARS_REGEX = /([a-zA-Z0-9@.])/;
  * @param {string} props.placeholder Input placeholder text
  */
 
-export const MultiInput = ({ placeholder }) => {
+export const MultiInput = ({
+  placeholder,
+  regex,
+  SelectionTemplate = DefaultSelectionTemplate,
```
Comment (aitchiss) [inline app/javascript/crayons/MultiInput/MultiInput.jsx:23]: (nitpick) can we add the new props to the JSDoc comments above?

File: app/javascript/crayons/MultiInput/MultiInput.jsx [hunk 3]
```diff
@@ -17,90 +17,219 @@ const ALLOWED_CHARS_REGEX = /([a-zA-Z0-9@.])/;
...
+        break;
+      case KEYS.DELETE:
+        if (currentValue === '') {
+          e.preventDefault();
+          editPreviousSelectionIfExists();
+        }
         break;
       default:
-        if (!ALLOWED_CHARS_REGEX.test(e.key)) {
+        if (!regex.test(e.key)) {
```
Comment (aitchiss) [inline app/javascript/crayons/MultiInput/MultiInput.jsx:85]: (nitpick) - I don't think we've noted `regex` as a required prop, but we also don't default its value of line 22 & expect it to exist here 🤔 

Maybe instead we should cover the scenario where no regex is provided and "anything goes":
```suggestion
        if (regex && !regex.test(e.key)) {
```

File: app/javascript/crayons/MultiInput/MultiInput.jsx [hunk 4]
```diff
@@ -17,90 +17,219 @@ const ALLOWED_CHARS_REGEX = /([a-zA-Z0-9@.])/;
...
+        aria-hidden="true"
+        className="absolute pointer-events-none opacity-0 p-2"
+      />
+
+      {/* A visually hidden list provides confirmation messages to screen reader users as an item is selected or removed */}
+      <div className="">
+        <p>Selected items:</p>
+        <ul
+          ref={selectedItemsRef}
+          className=" list-none"
```
Comment (aitchiss) [inline app/javascript/crayons/MultiInput/MultiInput.jsx:207]:
```suggestion
          className=" list-none screen-reader-only"
```
(not blocking) - not sure if we want to make this "hidden from view" here, or in a later PR

File: app/javascript/crayons/MultiInput/MultiInput.jsx (hunk 3 above)
Comment (Ridhwana) [reply app/javascript/crayons/MultiInput/MultiInput.jsx:85]: I like this, it makes a lot of sense, thanks for noting this!

**Comment by Ridhwana** [comment]
> We do have a lot of code in common between MultiSelectAutocomplete and here too but I feel like the mantra is "write everything twice" before creating abstractions so I guess just something we can keep our eyes on in future if we have need for these things in later components

I agree! I'm going to evaluate that at the end and make some notes and add it to our board so that I can tackle it during the next cooldown.

**Comment by aitchiss** [comment]
> Also, I couldn't help share my chuckle over this one love it the way loom changed "input" to "in poop". 😂

🤣  ahahahaha I'm going to be internally thinking about all inputs as in-poops now

**Pull request merged**